<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [45]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.vacancies
'''

In [5]:
# результат запроса
count_vacancies = pd.read_sql_query(query_3_1, connection)
display (count_vacancies)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
select count(id)
from public.employers
'''

In [7]:
# результат запроса
count_employers = pd.read_sql_query(query_3_2, connection)
display (count_employers)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select count(id)
from public.areas
'''


In [9]:
# результат запроса
count_areas = pd.read_sql_query(query_3_3, connection)
display (count_areas)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select count(id)
from public.industries
'''

In [11]:
# результат запроса
count_industries = pd.read_sql_query(query_3_4, connection)
display (count_industries)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

***

### выводы по предварительному анализу данных

В нашей базе данных 49197 уникальных вакансий  от  23501 работадателей из 1362 регионов . Работадатели представляют 294 сферы деятельности  

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
# Посмотрим первые 20 регионов и вычислим суммарное кол-во вакансий в них 
query_4_1 = f'''

(select count(v.id) cnt,
a.name area
from public.vacancies v
left join public.areas a on v.area_id = a.id
group by area
order by cnt desc
limit 20
)
union all

select sum (b.cnt) ,
'Total'

from (
select count(v.id) cnt,
a.name area
from public.vacancies v
left join public.areas a on v.area_id = a.id
group by area
order by cnt desc
limit 20 
) as b

'''

In [13]:
# результат запроса
top_areas = pd.read_sql_query(query_4_1, connection)
display (top_areas)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt             area
0    5333.0           Москва
1    2851.0  Санкт-Петербург
2    2112.0            Минск
3    2006.0      Новосибирск
4    1892.0           Алматы
5    1698.0     Екатеринбург
6    1670.0  Нижний Новгород
7    1415.0           Казань
8    1301.0        Краснодар
9    1144.0           Самара
10   1131.0   Ростов-на-Дону
11   1063.0          Воронеж
12    851.0          Ташкент
13    847.0       Красноярск
14    840.0       Нур-Султан
15    786.0        Челябинск
16    771.0            Пермь
17    767.0              Уфа
18    692.0            Томск
19    617.0             Омск
20  29787.0            Total

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
select count(id) cnt
from public.vacancies v
where v.salary_from is not Null or v.salary_to is not Null
'''

In [15]:
# результат запроса
vac_salary = pd.read_sql_query(query_4_2, connection)
display (vac_salary)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
select avg(salary_from) salary_from,
avg(salary_to) salary_to
from public.vacancies v
'''

In [17]:
# результат запроса
df_sal = pd.read_sql_query(query_4_3, connection)
display (df_sal)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


salary_from      salary_to
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
select schedule,
employment,
count (v.id) cnt
from public.vacancies v
group by schedule, employment
order by cnt desc
--limit 3
'''

In [19]:
# результат запроса
df_sh_em = pd.read_sql_query(query_4_4, connection)
display (df_sh_em)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''
select experience,
count (v.id) cnt
from public.vacancies v
group by experience
order by cnt
'''

In [21]:
# результат запроса
df_exp = pd.read_sql_query(query_4_5, connection)
display (df_exp)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

### выводы по детальному анализу вакансий
1. Более 60 % (около 29 000 вакансий ) состедоточены в 20 из 768 городах. Большинство вакансий из нашей БД предполагает работу  крупных городах миллионниках , таких как : Москва , Санкт-Петербург, Минск, Новосибирск, Алматы и д 
2. Так как хотя бы одно поле с з/п заполнено в менее чем 50% вакансий (24073 из 49197) можем сделать вывод что уровень з/п может сильно варьироваться в зависимости результатов собеседование
3. Уровень предлагаемой з/п в среднем от 71 до 110 тыс. руб. В сочетании с предыдущим пунктом можем отметить , что работодатели открывают для соискателей относительно невысокий уровень з/п. "Дорогие" з/п "не светят", хотя среднее значение не показатель , нужно смотреть график распределения.
4. Подавляющее болшинство вакансий вне зависимости от типа рабочего графика предполагает Полную занятость.
5. Около половины 26 152 вакансий для джуниор-специалистов (от 1 года до 3х). Около 14% вакансий 7197 без Опыта, что должно очень радовать таких как мы студентов 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:

# соединяем слева вакансии с работодателями по id работадателя
# группируем по имени работодателя и сортируем по кол-ву опубликованных работодателем вакансий в порядке возрастания
query_5_1 = f'''
select e.name,
count (v.id) cnt
from public.vacancies v
left join public.employers e on v.employer_id = e.id
group by e.name
order by cnt desc
limit 6
'''

In [23]:
# результат запроса
df_emp = pd.read_sql_query(query_5_1, connection)
display (df_emp)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331
5    ИК СИБИНТЕК   327

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [38]:
# присоединяем слева к регионам вакансии по id региона а после присоединяем работодателей по id региона работодателя
# группируем по региону , оставляем строки с нулевыми значениями (нет вакансий в регионе), сортируем 
query_5_2 = f'''
select a.name,
count (v.id) v_cnt,
count (e.id) e_cnt
from public.areas a
left join public.vacancies v on v.area_id = a.id
left join public.employers e on a.id = e.area
group by a.name
having count(v.id) = 0
order by e_cnt desc
'''

In [39]:
# результат запроса
df_area = pd.read_sql_query(query_5_2, connection)
display (df_area)

# Вообще с этим решением что-то не так. Если прописать distinct внутри count , то запрос не выполняеися , а если как все советуют having is null , вместо =0 ,
# то это вообще н7ичего не должно выводить, потому что при count null превращается в 0. или я не прав((( 


c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  v_cnt  e_cnt
0                          Россия      0    410
1                       Казахстан      0    207
2              Московская область      0     75
3              Краснодарский край      0     19
4              Ростовская область      0     18
..                            ...    ...    ...
588                      Струнино      0      0
589    Славгород (Алтайский край)      0      0
590                     Сорочинск      0      0
591  Ленинский (Тульская область)      0      0
592                         Тавда      0      0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# К работодателям слева присоединяем вакансии по id работодателя 
query_5_3 = f'''
select e.name,
count (distinct v.area_id) a_cnt

from public.employers e
left join public.vacancies v on v.employer_id = e.id
group by e.name
order by a_cnt desc
'''


In [27]:
# результат запроса
df_emp_area = pd.read_sql_query(query_5_3, connection)
display (df_emp_area)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  a_cnt
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
23170                  СДЕЛКА      0
23171            Alandr Group      0
23172                     СДК      0
23173      Lemon Land Lombard      0
23174                  Махина      0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''
select count( distinct e.id)
from  public.employers e
left join public.employers_industries ei on e.id = ei.employer_id 
where industry_id is null
'''

In [29]:
# результат запроса
df_null_ind = pd.read_sql_query(query_5_4, connection)
display (df_null_ind)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''
select e.name,
count(*)
from  public.employers e
left join public.employers_industries ei on e.id = ei.employer_id
group by  e.name
having count( distinct industry_id) = 4
order by name
limit 1
offset 2
'''

In [31]:
# результат запроса
df_4ind = pd.read_sql_query(query_5_5, connection)
display (df_4ind)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
select count( distinct e.id)
from  public.employers e
left join public.employers_industries ei on e.id = ei.employer_id
join industries i on i.id=ei.industry_id
where i.name ='Разработка программного обеспечения'
'''

In [33]:
# результат запроса
df_ind_name = pd.read_sql_query(query_5_6, connection)
display (df_ind_name)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников

import urllib.parse as req
import requests

# создаем переменную  url с адресом страницы на Вики
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

# создаем список из городов первой таблицы (столбец Город) что прочитает метод read_html из адреса url
cities = list(pd.read_html(url)[1]['Город'])
cities[-3] = 'Воронеж' # Исправляем недочет с некоректным чтением города Воронеж
cities = tuple(cities) # для sql запроса приготовим кортедж 
cities


('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [35]:
# К таблице с регионами присоединяем вакансии и работодателей
# Устанавливаем условие работодатель = Яндексу , группируем и фильтруем только регионы из кортеджа cities
# Во второй части используем union для формирования строчки Total
query_5_7 = f'''
select a.name,
count(v.id) cnt
from public.areas a
join public.vacancies v on v.area_id = a.id
join public.employers e on v.employer_id = e.id
where e.name = 'Яндекс'
group by a.name
having a.name in {cities}

union all
select 
'Total',
sum(b.cnt)
from 
(
select a.name,
count(v.id) cnt
from public.areas a
join public.vacancies v on v.area_id = a.id
join public.employers e on v.employer_id = e.id
where e.name = 'Яндекс'
group by a.name
having a.name in {cities}
) as b
order by cnt
'''

In [36]:
# результат запроса
df_ya = pd.read_sql_query(query_5_7, connection)
display (df_ya)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

### выводы по анализу работодателей
1. Крупнейшие работодатели - ИТ-компании, Банки, Нефтедобывающие компании . На первом месте по кол-ву вакансий - Яндекс 1933 вакансии.
2. Лидеры по кол-ву регионов в которых опубликованы вакансии: Яндекс , Ростелеком , Спецремонт
3. 8419 работадателенй не указали сферу деятельности , это около 40 % 
4. Разработкой ПО занимаются 3553 компании
5. Яндекс разместил 485 вакансий в городах миллионниках. Это менее трети всех вакансий, размещенных Яндексом , что подтверждает его лидерство по охвату городов , а та кже говорит о том что даже в небольших городах можно найти вакансию этой компании

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса
query_6_1 = f'''
select count(v.name)
from  public.vacancies v

where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [ ]:

# результат запроса
df_data = pd.read_sql_query(query_6_1, connection)
display (df_data)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [48]:
# делаем выборку сначала по названиям вакансий для дата-сайтистов , 
# по условию 'and' добавляем к выборке "начинающего"
# для исключения ошибки регистра используем lower чтобы привести все строки к нижнему регистру 
query_6_2 = f'''
select count(v.id)

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and
(
lower(v.name) like '%junior%' or
v.experience like '%Нет опыта%' or
v.employment like '%Стажировка%'
)

'''

In [49]:
# результат запроса
df_ds = pd.read_sql_query(query_6_2, connection)
display (df_ds)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# в этом запросе используем фильтраци по имени для DS и по  key_skills для sql и postgres
query_6_3 = f'''
select count(v.id)

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and
(
lower(v.key_skills) like '%sql%' or
lower(v.key_skills) like '%postgres%'
)

'''

In [44]:
# результат запроса
df_sql = pd.read_sql_query(query_6_3, connection)
display (df_sql)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = f'''
select count(v.id)

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and
(
lower(v.key_skills) like '%python%' 
)

'''

In [ ]:
# результат запроса
df_py = pd.read_sql_query(query_6_4, connection)
display (df_py)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# для запроса используем разницу между длиной строки key_skills и длиной этой же строки в которой убираем (заменяем на '') символы '/t' 
# учитывая что колличество навыков описаных в строке key_skills на 1 больше чем разделителей к результату предыдущей операции прибавляем единицу

query_6_5 = f'''
select 
avg (char_length( v.key_skills)- char_length(replace (v.key_skills, CHR(9),'' ))+1 )   avg

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
) 
and v.key_skills is not null

'''

In [ ]:
# результат запроса
df_ds_skill = pd.read_sql_query(query_6_5, connection)
display (df_ds_skill)

                    


c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg
0  6.406032

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# алгоритм логики запроса следующий:
# 1. делаем выборку [опыт , средняя зарплата], в котрой ср. з/п считаем как сумму (salary_from + salary_to) при условии что одна из них нулевая 
# чтобы превратить null в  0  используем coalesce
# 2. делаем выборку [опыт , средняя зарплата], в котрой ср. з/п считаем как сумму деленную на 2 (salary_from + salary_to)/2 при условии что оба значение not null
# 3. C помощью union соединяем результат 1 и 2 
# 4. Группируем по столбцу  опыт 
query_6_6 = f'''
select 
b.experience ,
avg(b.sallary)
from 
(
select 
v.experience experience,
coalesce(v.salary_from, 0) + coalesce(v.salary_to, 0) sallary
from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
) 

and 

(
    (v.salary_from is null and v.salary_to is not null)
    or (v.salary_from is not null and v.salary_to is  null)
)


union all

select 
v.experience,
(v.salary_from + v.salary_to)/2
from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
) 

and 

(
    (v.salary_from is not null and v.salary_to is not null)
 
)


) as b

group by 
b.experience

'''

In [ ]:
# результат запроса
df_ds_avg = pd.read_sql_query(query_6_6, connection)
display (df_ds_avg)


c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience            avg
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  139674.750000
2       От 3 до 6 лет  243114.666667

***

### выводы по предметному анализу
1.  1771 вакансия имеет отношение к данным
2. Для начинающего дата сайнтиста найдется 61 вакансия в том числе без опыта и на стажировку. Кстати , это довольно большой процент если учесть что вакансий всего DS вакансий 480 
3.  Около половины вакансий по DS - 201 из 480 трубуют занния SQL или Postgres , это говорит о высокой важности навыка при отборе кандидатов 
4. 351 дата-сайтистов должны знать язык программирования python. интересно посмотреьть на остальные , какой же там язык. Сделаем доп исследование.
5. Более 6 ключевых навыков должно быть у претендентов на вакансию DS. Это говорит о том что работа требует больших знаний в разных областях 
6. Чем выше опыт тем выше з/п DS . 

# Общий вывод по проекту

In [79]:
# проанализируем требования к DS помимо Python и SQL . 
# Сначала посмотрим на общую картину 
query_7_1 = f'''
select *

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
limit 40

'''


In [80]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [81]:
df_add_skill = pd.read_sql_query(query_7_1, connection)
pd.set_option('display.max_rows', None)
display (df_add_skill)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


id                                               name  \
0   50533416                            Junior+ CV/ML developer   
1   54824572                                     Data Scientist   
2   54825318                           Data Scientist (Junior+)   
3   66120847                              AI/ML Senior Engineer   
4   55351147                                     Data Scientist   
5   55351168                                     Data Scientist   
6   55351179                                     Data Scientist   
7   55351195                                     Data Scientist   
8   66120842                              AI/ML Senior Engineer   
9   55166994                          ML разработчик (удалённо)   
10  55373225                             Разработчик C++/Qt/QML   
11  55092666                                        ML Engineer   
12  52995746                              C++ разработчик (QML)   
13  55299643                       Риск-менеджер/Data Scientist   
14  55553897                     Junior Data Scientist (Стажер)   
15  55529384     Математик, Junior/Middle/Senior Data Scientist   
16  55204390                              Middle Data Scientist   
17  55145641                              Junior Data Scientist   
18  44095047     Инженер-тестировщик в команду Machine Learning   
19  55367723                           Data Scientist/Математик   
20  55343709  Data Scientist/Machine Learning Engineer (Deve...   
21  55108462                         ML Engineer/Data Scientist   
22  55499148                        Специалист Data Science (R)   
23  55313995    C++ Machine Learning Engineer (Computer Vision)   
24  54659098                    Data Scientist / ML-разработчик   
25  54475492                            Data Scientist (Junior)   
26  54324720                      Leader of ML Engineering Team   
27  55338837                              ML разработчик в Дзен   
28  55375909                                     Data Scientist   
29  55119989                       Intern/Junior Data Scientist   
30  52494878             Python разработчик / Машинное обучение   
31  54592046                       Data Scientist/Data Engineer   
32  55379599                                     Data Scientist   
33  53545373  Разработчик Python ML/обработка естественного ...   
34  55120211                         Junior Data Scientist (VK)   
35  55120401                       Junior/Middle Data Scientist   
36  55120848                                     Data Scientist   
37  55120849                                     Data Scientist   
38  55120850                                     Data Scientist   
39  55120851                                     Data Scientist   

                                           key_skills          schedule  \
0           Python\tSQL\tРабота в команде\tOpenCV\tML     Гибкий график   
1   Oracle Pl/SQL\tБазы данных\tMachine Learning\t...       Полный день   
2   Python\tOpenCV\tTensorflow\tComputer Vision\tН...       Полный день   
3                      Python\tFlask\tLinux\tAWS\tGit     Гибкий график   
4    Python\tSQL\tNLP\tScikit-learn\tMachine Learning       Полный день   
5    Python\tSQL\tNLP\tScikit-learn\tMachine Learning       Полный день   
6    Python\tSQL\tNLP\tScikit-learn\tMachine Learning       Полный день   
7    Python\tSQL\tNLP\tScikit-learn\tMachine Learning       Полный день   
8                      Python\tFlask\tLinux\tAWS\tGit     Гибкий график   
9   Python\tПоиск информации в интернет\tC++\tМате...  Удаленная работа   
10                         C++\tQt\tVoIP\tTCP/IP\tQML       Полный день   
11                           Python\tGit\tLinux\tBash       Полный день   
12                                      ООП\tC++\tSQL       Полный день   
13     Python\tSQL\tSas\tMS PowerPoint\tАнализ данных       Полный день   
14  Python\tSQL\tData Mining\tC++\tМатематический ...     Гибкий график   
15  Python\tMATLAB\tМатематическое моделирование\t...  Удаленн

In [117]:
# 1. Визуальный Анализ оставшихся 52 вакансий показывает , что при анализе DS мы не учли , что нужно исключить QML - не имеет отношение к DS , добавим это условие к фильтрации
# 2. Выведем таблицу с другими ключевыми навыками и их кол-вом для DS 
query_7_2 = f'''
select 'Oracle' skill,
count(*) cnt

from   public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%' and  v.name not like '%QML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and 
(
  lower(v.key_skills) like '%oracle%'  
)

union

select 'git',
count(*) cnt

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%' and  v.name not like '%QML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and 
(
  lower(v.key_skills) like '%git%'  
)

union

select 'unix or linux',
count(*) cnt

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%' and  v.name not like '%QML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and 
(
  lower(v.key_skills) like '%unix%' or  lower(v.key_skills) like '%linux%'
)

union

select 'C++',
count(*) cnt

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%' and  v.name not like '%QML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and 
(
  lower(v.key_skills) like '%c++%'
)


union

select 'Математическая статистика',
count(*) cnt

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%' and  v.name not like '%QML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and 
(
  lower(v.key_skills) like '%математическая статистика%'
)

union

select 'numpy',
count(*) cnt

from  public.vacancies v

where 
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
(v.name like '%ML%' and v.name not like '%HTML%' and  v.name not like '%QML%') or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%'
)
and 
(
  lower(v.key_skills) like '%numpy%'
)

order by cnt
'''

In [118]:
df_add2_skill = pd.read_sql_query(query_7_2, connection)
display (df_add2_skill)

c:\Users\Андрей\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


skill  cnt
0                     Oracle   11
1                        C++   32
2                      numpy   41
3              unix or linux   48
4  Математическая статистика   62
5                        git   68

### Итог Исследования

1. Рынок вакансий очень обширный и работу сможет найти как специалист без опыта , так и с опытом. Более 14 % предполагают отсутсвие опыта или стажировку
2. Стоимсоть специалистов на рынке сильно зависит от навыков и результатов собеседований . Более половины работодателей не публикуют предлагаемую з/п
3. Крупнейшие компании размещают свои вакансии не только в крупных мегаполисах , но и небольших населенных пунктах . Например Яндекс разместил только треть своих вакансий в городах миллионниках 
4. Вакансии имеющие отношения к данным занимают около 4% от всей БД. Вакансии по DS  имеют опубликованную з/п от 74 000 до 241 000 рублей. 
5. Специальность DS предполагает большой объем знаний в различных областях . В перечне требуемых навыков не менее 6-ти пунктов по этим вакансиям.
6. Дополнительное исследование выявило ошибку в фильтрации вакансий DS. Необходимо исключить QML
7. Список популярных требуемых скилов для DS : Git, Математическая статистика, unix or linux, numpy, C++, Oracle